Louis/bgp testing

ldelossa opened the pull request:
This PR introduces the necessary commits for testing Cilium's BGP Speaker and BGP Manager in isolation. 

A new interface is built which abstracts the MetalLB code slightly. This will allow unit tests to instantiate Cilium's speaker and manager with a mock. 

Each mock will be able to record the invocation of their methods, useful in proving the Speaker  and Manager are using the MetalLB components correctly. 

This PR will merge back into "pr/christarazi/podcidrs-bgp" topic branch and NOT master.

christarazi commented on pkg/bgp/speaker.go:
```diff
@@ -0,0 +1,96 @@
+package bgp
```
Copyright header at the top. You can simply copy-pasta from another file

christarazi commented on pkg/bgp/speaker.go:
```diff
@@ -0,0 +1,96 @@
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
```
Our preferred ordering (with newlines in between) for imports is:

```
stdlib

cilium pkgs

third party pkgs
```

ldelossa replied:
Hmm, pretty sure "gopls" is popping these into the order shown. Let me see if i can manually reorder and gopls not move them around.

ldelossa replied:
Cool, gopls allows for this:

```
	"context"
	"fmt"
	"os"

	bgpconfig "github.com/cilium/cilium/pkg/bgp/config"
	bgpk8s "github.com/cilium/cilium/pkg/bgp/k8s"
	bgplog "github.com/cilium/cilium/pkg/bgp/log"
	"github.com/cilium/cilium/pkg/logging"
	"github.com/cilium/cilium/pkg/logging/logfields"
	nodetypes "github.com/cilium/cilium/pkg/node/types"
	"github.com/cilium/cilium/pkg/option"

	"go.universe.tf/metallb/pkg/k8s/types"
	metallbspr "go.universe.tf/metallb/pkg/speaker"
```

look cool?

christarazi commented on pkg/bgp/speaker.go:
```diff
@@ -0,0 +1,96 @@
+	return m.C.SetNodeLabels(m.logger, labels)
+}
+func (m MetalLBSpeaker) PeerSessions() []metallbspr.Session {
+	return m.C.PeerSessions()
```
These should probably be pointers because when creating an object to adhere to the `Speaker` interface, it's going to need to be a pointer anyway. Any reason why they weren't initially?

```suggestion
func (m *MetalLBSpeaker) SetService(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState {
	return m.C.SetService(m.logger, name, svc, eps)
}
func (m *MetalLBSpeaker) SetNodeLabels(labels map[string]string) types.SyncState {
	return m.C.SetNodeLabels(m.logger, labels)
}
func (m *MetalLBSpeaker) PeerSessions() []metallbspr.Session {
	return m.C.PeerSessions()
```

ldelossa replied:
Typically, when all members of a struct are pointers, I pass the container object by value. The container struct doesn't need to add to gc pressure (by being a heap allocation on every construction). 

I typically default to copying over passing by reference until i can measure and know copying is a problem. 

Im not immediately seeing the issue with the receiver being a pointer or a value, since the constructor returns the interface type.

christarazi commented on pkg/bgp/speaker.go:
```diff
@@ -0,0 +1,96 @@
+		return nil, err
+	}
+	if ss := c.SetConfig(logger, config); ss == types.SyncStateError {
+		return nil, fmt.Errorf("Setting MetalLB config failed. View logs for details")
```
Errors should be lower-cased: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

[1]

christarazi replied:
Additionally, this error msg borders on being a log msg. Maybe we can simply say 
```suggestion
		return nil, fmt.Errorf("failed to set MetalLB config")
```
and the caller can customize the log msg how they please (if this error was converted into a static var). Lower level error msgs should be simpler and the higher you go up the stack, the more descriptive they can become as they stack, etc. WDYT?

ldelossa replied:
I agree, does look like a log.

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -45,51 +42,25 @@ var _ subscriber.Node = (*Speaker)(nil)
+	ctrl, err := bgp.NewMetalLBSpeaker(ctx)
 	if err != nil {
-		log.WithError(err).Fatal("Failed to parse BGP configuration")
+		return nil, fmt.Errorf("Could not create MetalLB speaker: %v", err)
```
[1]

christarazi reviewed (commented):
At a high-level I am wondering why not replace the `pkg/bgp/speaker/speaker.go` implementation with the one implemented here as `pkg/bgp/speaker.go`. 

A few comments to address below as well.

ldelossa committed:
bgp: rename Speaker to MetalLBSpeaker

this commit renames our current bgp Speaker struct to MetalLBSpeaker as
preparation for further refactoring to support unit testing.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp: introduce speaker interface and metallb implementation

this commit creates a Speaker interface allowing a level of indirection
between the MetalLBSpeaker event handling and the calling of BGP
specific methods.

this is useful for isolating the MetalLBSpeaker business logic for unit
testing.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp: refactor MetalLBSpeaker to use Speaker interface

This commit wires up the MetalLBSpeaker struct to utilize the new
Speaker interface and MetalLB driven implementation.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp: introduce bgp controller interface

this commit introduces an interface for abstracting the use of the
metallb controller within the bgp Manager.

this allows for isolated testing of the bgp Manager.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp: wire up bgp Controller interface to bgp Manager

this commit refactors the necessary code to plumb the bgp Controller
interface into the bgp Manager.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

christarazi commented on pkg/bgp/manager/metallb.go:
```diff
@@ -0,0 +1,82 @@
+	metallbctl "go.universe.tf/metallb/pkg/controller"
+	"go.universe.tf/metallb/pkg/k8s"
+	"go.universe.tf/metallb/pkg/k8s/types"
+	v1 "k8s.io/api/core/v1"
```
Import ordering.

(Spent a few mins looking for tools that can auto detect this. Unfortunately, no tool exists that can detect our order of preference and auto group them together. :disappointed:)

ldelossa replied:
Not a problem. Will fix

christarazi commented on pkg/bgp/speaker/metallb.go:
```diff
@@ -0,0 +1,107 @@
+		return nil, err
+	}
+	if ss := c.SetConfig(logger, config); ss == types.SyncStateError {
+		return nil, fmt.Errorf("Setting MetalLB config failed. View logs for details")
```
This error msg hasn't been addressed. Looks like a log msg

christarazi commented on pkg/bgp/speaker/pod_cidr.go:
```diff
@@ -30,10 +30,10 @@ var (
-	for _, session := range s.PeerSessions() {
+	for _, session := range s.speaker.PeerSessions() {
 		go func(sess metallbspr.Session) { // Need an outer closure to capture session.
 			wg.Add(1)
```
There are a few go lint warnings here that we should fix in the original PR, FYI.

ldelossa replied:
Cool, I'll make note to lint before final merger.
